packages/gatsby/src/datastore/common/iterable.ts

```diff
@@ -1,44 +1,153 @@
-import { IGatsbyIterable } from "../types"
-
-export class GatsbyIterable<T> implements IGatsbyIterable<T> {
-  constructor(private source: Iterator<T>) {}
+/**
+ * Wrapper for any iterable providing chainable interface and convenience methods
+ * similar to array.
+ *
+ * Additionally provides convenience methods for sorted iterables.
+ *
+ * Note: avoiding async iterables because of perf reasons, see https://github.com/nodejs/node/issues/31979
+ * (fortunately lmdb can traverse stuff in sync manner very fast)
+ */
+export class GatsbyIterable<T> {
+  constructor(private source: Iterable<T> | (() => Iterable<T>)) {}
 
   [Symbol.iterator](): Iterator<T> {
-    return this.source
+    const source =
+      typeof this.source === `function` ? this.source() : this.source
+    return source[Symbol.iterator]()
   }
 
-  concat<U>(other: Iterable<U>): GatsbyIterable<T | U> {
-    return new GatsbyIterable(concatSequence(this, other))
+  concat<U = T>(other: Iterable<U>): GatsbyIterable<T | U> {
```

Comment:
Does setting this default make any difference? Won't `U` be inferred from type of argument (`other: Iterable<U>`)?

Reply:
Yup, seems redundant. Don't remember why I've added this 🤷‍♂️  Removed.